src/tree/gpu_hist/histogram.cuh

```diff
@@ -0,0 +1,40 @@
+/*!
+ * Copyright 2020 by XGBoost Contributors
+ */
+#ifndef HISTOGRAM_CUH_
+#define HISTOGRAM_CUH_
+#include <thrust/transform.h>
+#include "../../data/ellpack_page.cuh"
+
+namespace xgboost {
+namespace tree {
+
+GradientPair CreateRoundingFactor(common::Span<GradientPair> gpair);
+
+DEV_INLINE float TruncateWithRoundingFactor(float rounding_factor, float x) {
```

Comment:
I don't understand how this function could be so expensive that it slows down the algorithm if we use it in the histogram kernel.

Reply:
@RAMitchell I pushed the change for computing truncation on the fly.  There are other functions like `GetHess`, `GetGrad` and constructors.  I'm not sure if the compiler can optimize them away.